tools/validators/instance_validator/universe_generation/generate_universe.py

```diff
@@ -0,0 +1,213 @@
+    field_folders = parse.ParseFieldFoldersFromFiles(config.fields,
+                                                     subfield_universe,
+                                                     state_universe)
+    field_universe = field_lib.FieldUniverse(field_folders)
+
+  return field_universe
+
+def build_type_universe(config, field_universe):
+  """Builds TypeUniverse object.
+
+  Args:
+    config: a Config namedtuple containing lists of localpaths to config files.
+    field_universe: a FieldUniverse object to validate the TypeUniverse.
+
+  Returns:
+     A TypeUniverse that is fully populated with type content specified
+     in the config.
+  """
+
+  type_folders = parse.ParseTypeFoldersFromFiles(config.type_defs,
+                                                 field_universe)
+  type_universe = entity_type_lib.EntityTypeUniverse(type_folders)
+
+  return type_universe
+
+def build_universe():
+  """Generates the ontology universe.
+
+  Returns:
+    Generated universe object.
+  """
+
+  ontology_validator_exists = path.exists(path.join(
+      '..', 'ontology_validator'))
+  ontology_exists = path.exists(path.join('..', '..', '..', 'ontology'))
+
+  if not (ontology_validator_exists and ontology_exists):
+    print('ERROR: ontology validator or ontology have changed locations')
+    return None
+
+  yaml_files = external_file_lib._RecursiveDirWalk(path.join(
+      '..', '..', '..', 'ontology', 'yaml', 'resources'))
+  config = presubmit_validate_types_lib.SeparateConfigFiles(yaml_files)
+  universe = presubmit_validate_types_lib.BuildUniverse(config)
+
+  namespace_validation = namespace_validator.NamespaceValidator(
+      universe.GetEntityTypeNamespaces())
+
+  if not namespace_validation.IsValid():
+    return None
+
+  return universe
+
+def parse_universe(universe):
+  """Parses generated the ontology universe object into its subcomponents
+
+  Args:
+    universe: generated the ontology universe object
+
+  Returns:
+    fields: valid universe field types generated from the ontology
+    subfields_map: valid universe subfield types generated from the ontology
+    states_map: valid universe state types generated from the ontology
+    units_map:  valid universe unit types generated from the ontology
+    entities_map:  valid universe entity types generated from the ontology
+  """
+  states = universe.state_universe
+  entities = universe.entity_type_universe
+  units = universe.unit_universe
+  subfields = universe.subfield_universe
+
+  # USAGE: fields.IsFieldDefined('process_return_water_temperature_sensor', '')
+  fields = universe.field_universe
+
+  # consolidate all entity information into dictionary
+  entities_map = {}
+  entity_type_namespaces = entities.type_namespaces_map
+  for namespace in entity_type_namespaces.keys():
+    valid_types_map = entity_type_namespaces[namespace].valid_types_map
+
+    namespace_name = entity_type_namespaces[namespace].namespace
+    entities_map[namespace_name] = {}
```

Comment:
Why duplicate this information? It is already mapped in the universe, you can simply update the universe with a method to retrieve it.

Reply:
Removed